vscode-trace-extension/package.json

```diff
@@ -124,43 +129,50 @@
     },
     "views": {
       "trace-explorer": [
+        {
```

Comment:
I have some general comments/concerns regarding this approach. I recently raised an issue (https://github.com/eclipse-cdt-cloud/theia-trace-extension/issues/1053) to handle single trace files along with trace folders. With that, i was hoping to achieve two things:

1) Have a separate button to handle trace files (this is simple to achieve with your approach)

2) Extensions have the ability to render either one or both of the buttons - this is useful for cases when users/extending extensions dont have a use case for one of the buttons (e.g. our users only ever need to inspect files and never folders) -- I am not sure if that is possible with this configuration as far as i know. Maybe it is possible to have different welcome views based on when clause context?

Reply:
> I have some general comments/concerns regarding this approach. I recently raised an issue ([eclipse-cdt-cloud/theia-trace-extension#1053](https://github.com/eclipse-cdt-cloud/theia-trace-extension/issues/1053)) to handle single trace files along with trace folders. With that, i was hoping to achieve two things:
> 
> 1. Have a separate button to handle trace files (this is simple to achieve with your approach)
> 2. Extensions have the ability to render either one or both of the buttons - this is useful for cases when users/extending extensions dont have a use case for one of the buttons (e.g. our users only ever need to inspect files and never folders) -- I am not sure if that is possible with this configuration as far as i know. Maybe it is possible to have different welcome views based on when clause context?

1. I've added this functionality to this PR in a separate commit.  I think there is value in having this available by default, so just put it into the default view.
2. See this comment about modifying the contents of an extension's packate.json post-build: https://github.com/eclipse-cdt-cloud/vscode-trace-extension/pull/216#discussion_r1520809823.  I'm not sure if this helps you with your use-case, but its currently how I modify contents in order to suit custom needs.

Reply:
> 
>     2. Extensions have the ability to render either one or both of the buttons - this is useful for cases when users/extending extensions dont have a use case for one of the buttons (e.g. our users only ever need to inspect files and never folders) -- I am not sure if that is possible with this configuration as far as i know. Maybe it is possible to have different welcome views based on when clause context?

It's possible. In the packages.json you can have to entries for open directory and open files. Then there can be context flag added in the when-clause to show it or not. 
```json
"viewsWelcome": [
      {
        "view": "welcome",
        "name": "Open Trace Directory",
        "contents": "There are currently no opened traces.\n[Open Trace Folder](command:openedTraces.openTraceFolder)",
        "when": "trace-explorer.noExperiments || !traceViewer.serverUp"
      },
      {
        "view": "welcome",
        "name": "Open Trace File",
        "contents": "There are currently no opened traces.\n[Open Trace File](command:openedTraces.openTraceFile)",
        "when": "(trace-explorer.noExperiments || !traceViewer.serverUp) && trace-explorer.hasFileOpen"
      },      
      {
        "view": "welcome",
        "name": "Documentation",
        "contents": "To learn more about how to use the Trace Viewer [read our documentation](https://github.com/eclipse-cdt-cloud/vscode-trace-extension).",
        "when": "trace-explorer.noExperiments || !traceViewer.serverUp"
      }
    ],
```

This flag can be set programmatically from any extension.
```javascript
vscode.commands.executeCommand('setContext', 'trace-explorer.hasFileOpen', true);
```

@ngondalia  is that what you were looking for?

Reply:
> > 1. Have a separate button to handle trace files (this is simple to achieve with your approach)
> 
> I've added this functionality to this PR in a separate commit. I think there is value in having this available by default, so just put it into the default view.
> 
> > 2. Extensions have the ability to render either one or both of the buttons - this is useful for cases when users/extending extensions dont have a use case for one of the buttons (e.g. our users only ever need to inspect files and never folders) -- I am not sure if that is possible with this configuration as far as i know. Maybe it is possible to have different welcome views based on when clause context?
> 
> See this comment about modifying the contents of an extension's packate.json post-build: [#216 (comment)](https://github.com/eclipse-cdt-cloud/vscode-trace-extension/pull/216#discussion_r1520809823). I'm not sure if this helps you with your use-case, but its currently how I modify contents in order to suit custom needs.

- That wouldn't work for our case becasue we are not publishing our own version of the extension - we will be relying on the marketplace version, so its more preferrable to do it through API/contexts

Reply:
> > ```
> > 2. Extensions have the ability to render either one or both of the buttons - this is useful for cases when users/extending extensions dont have a use case for one of the buttons (e.g. our users only ever need to inspect files and never folders) -- I am not sure if that is possible with this configuration as far as i know. Maybe it is possible to have different welcome views based on when clause context?
> > ```
> 
> It's possible. In the packages.json you can have to entries for open directory and open files. Then there can be context flag (`hasFileOpen`) added that is checked in the when-clause. The default value would be `true`.
> 
> ```json
> "viewsWelcome": [
>       {
>         "view": "welcome",
>         "name": "Open Trace Directory",
>         "contents": "There are currently no opened traces.\n[Open Trace Folder](command:openedTraces.openTraceFolder)",
>         "when": "trace-explorer.noExperiments || !traceViewer.serverUp"
>       },
>       {
>         "view": "welcome",
>         "name": "Open Trace File",
>         "contents": "There are currently no opened traces.\n[Open Trace File](command:openedTraces.openTraceFile)",
>         "when": "(trace-explorer.noExperiments || !traceViewer.serverUp) && trace-explorer.hasFileOpen"
>       },      
>       {
>         "view": "welcome",
>         "name": "Documentation",
>         "contents": "To learn more about how to use the Trace Viewer [read our documentation](https://github.com/eclipse-cdt-cloud/vscode-trace-extension).",
>         "when": "trace-explorer.noExperiments || !traceViewer.serverUp"
>       }
>     ],
> ```
> 
> This flag can be set programmatically from any extension.
> 
> ```js
> vscode.commands.executeCommand('setContext', 'trace-explorer.hasFileOpen', false);
> ```
> 
> @ngondalia is that what you were looking for?

- Yeah, that is exactly what i was thinking! The only thing i would suggest is to break it down further to split the two buttons and that can have its own when context clauses:

```
"viewsWelcome": [
      {
        "view": "welcome",
        "name": "Open Trace Directory",
        "contents": "There are currently no opened traces.\n[Open Trace Folder](command:openedTraces.openTraceFolder)",
        "when": "trace-explorer.noExperiments || !traceViewer.serverUp"
      },
      {
        "view": "welcome",
        "name": "Open Trace File",
        "contents": "There are currently no opened traces.\n[Open Trace File](command:openedTraces.openTraceFile)",
        "when": "(trace-explorer.noExperiments || !traceViewer.serverUp) && trace-explorer.hasFileOpen"
      },      
      {
        "view": "welcome",
        "name": "Documentation",
        "contents": "To learn more about how to use the Trace Viewer [read our documentation](https://github.com/eclipse-cdt-cloud/vscode-trace-extension).",
        "when": "trace-explorer.noExperiments || !traceViewer.serverUp"
      }
    ]
```